hamilton/driver.py

```diff
@@ -840,6 +854,290 @@ def visualize_path_between(
         except ImportError as e:
             logger.warning(f"Unable to import {e}", exc_info=True)
 
+    def materialize(
+        self,
+        *materializers: materialization.MaterializerFactory,
+        additional_vars: List[Union[str, Callable, Variable]],
+        overrides: Dict[str, Any] = None,
+        inputs: Dict[str, Any] = None,
+    ) -> Tuple[Any, Dict[str, Any]]:
+        """Executes and materializes with ad-hoc materializers.
+        This does the following:
+            1. Creates a new graph, appending the desired materialization nodes
+            2. Runs the graph with the materialization nodes outputted, as well as any additional
+                nodes requested (which can be empty)
+            3. Returns a Tuple[Materialization metadata, additional_vars result]
+
+        For instance, say you want to Materialize the output of a node to CSV:
+
+        .. code-block:: python
+             from hamilton import driver, base
+             from hamilton.io.materialize import to
+             dr = driver.Driver(my_module, {})
+             # foo, bar are pd.Series
+             metadata, result = dr.Materialize(
+                 to.csv(
+                     path="./output.csv"
+                     id="foo_bar_csv",
+                     dependencies=["foo", "bar"],
+                     combine=base.PandasDataFrameResult()
+                 ),
+                 additional_vars=["foo", "bar"]
+             )
+
+        The code above will Materialize the dataframe with "foo" and "bar" as columns, saving it as
+        a CSV file at "./output.csv". The metadata will contain any additional relevant information,
+        and result will be a dictionary with the keys "foo" and "bar" containing the original data.
+
+        Note that we pass in a `ResultBuilder` as the `combine` argument, as we may be materializing
+        several nodes.
+
+        additional_vars is used for debugging -- E.G. if you want to both realize side-effects and
+        return an output for inspection. If left out, it will return an empty dictionary.
+
+        If you don't want to join results (and are instead outputting a single value), you can
+        bypass the `combine` keyword.
+
+        .. code-block:: python
+             from hamilton import driver, base
+             from hamilton.io.materialize import to
+             dr = driver.Driver(my_module, {})
+             # foo, bar are pd.Series
+             metadata, _ = dr.Materialize(
+                 to.csv(
+                     path="./output.csv"
+                     id="foo_bar_csv",
+                     dependencies=["foo_bar_already_joined],
+                 ),
+             )
+
+        This will just save it to a csv.
+
+        Note that materializers can be any valid DataSaver -- these have an isomorphic relationship
```

Comment:
should link ideally to DataSaver reference ideally...

Reply:
Going to do a quick documentation push, separately, prior to release. This will have the data saver/loader framework.